### ISSUE TITLE
Fundamental difference between fast cover and cover

### ISSUE THREAD

**Issue by iamrohitbanga** [facebook/zstd#1654]
Hello

I am trying to understand the code for cover and fastcover dictionary training techniques. I understand from reading other github issues and the code, that cover is expensive. e.g. it [consumes ~ 11x the memory of samples](https://github.com/facebook/zstd/issues/513#issuecomment-273358245). Fast cover is optimized and has a smaller footprint.

What is the fundamental algorithmic change from cover to fastcover? How does it achieve the trade-off between accuracy and speed and what is the thing that slows down the old cover implementation?

I see that the training algorithm is based on the paper "Effective Construction of Relative Lempel-Ziv Dictionaries". Does fastcover have some document to understand the conceptual difference?

Looking at the code it seems FastCover computes a hash function on the dmers. The other difference is the absence of COVER_group() function and suffix properties in FASTCOVER_ctx_t.

Thanks

**Comment by terrelln** [comment]
They are using the exact same algorithm. Fast cover just uses some approximations to save CPU and memory. The difference in dictionary effectiveness between the two shouldn't be more than a percent or two most of the time.

Instead of [building a partial-suffix array](https://github.com/facebook/zstd/blob/dev/lib/dictBuilder/cover.c#L611) and computing frequencies exactly, and using a hash table to keep track of the frequencies in a segment, we hash the dmers, and compute the frequencies of the hash. This avoids an expensive sort. Additionally, computing frequencies of the hash means we don't have to deal with collisions, since we treat colliding dmers as the same. We can tune the size of the hash, larger means better dictionary but slower, smaller means faster construction, but maybe a worse dictionary.

There are a few other smaller optimizations:
* Fast cover optionally [skips over the input](https://github.com/facebook/zstd/blob/dev/lib/dictBuilder/fastcover.c#L265) when computing frequencies instead of looking at every position
* We don't pass all of the samples to `ZDICT_finalizeDictionary()`
* We can use a larger test set, and a smaller training set
* We can try less parameters during parameter optimizations
